src/WasmModule.ts

```diff
@@ -0,0 +1,193 @@
+import { decode } from '@webassemblyjs/wasm-parser';
+import ast from '@webassemblyjs/ast';
+
+import Graph from './Graph';
+import { IdMap, ModuleJSON, WebAssemblyJSAst } from './rollup/types';
+
+import { makeLegal } from './utils/identifierHelpers';
+import { RenderOptions } from './utils/renderHelpers';
+import { basename, extname } from './utils/path';
+
+import ExternalModule from './ExternalModule';
+import Module from './Module';
+
+import ModuleScope from './ast/scopes/ModuleScope';
+import Variable from './ast/variables/Variable';
+import NamespaceVariable from './ast/variables/NamespaceVariable';
+
+const decoderOpts = {
+	ignoreCodeSection: true,
+	ignoreDataSection: true
+};
+
+const buildLoader = ({ URL, IMPORT_OBJECT }) => `
+	function then(resolve) {
+		if (typeof WebAssembly.instantiateStreaming !== 'function') {
+		  throw new Error('WebAssembly.instantiateStreaming is not supported');
+		}
+
+		if (typeof window.fetch !== 'function') {
+		  throw new Error('window.fetch is not supported');
+		}
+
+		const req = window.fetch('${URL}');
+
+		WebAssembly
+			.instantiateStreaming(req, ${IMPORT_OBJECT || '{}'})
+			.then(res => res.instance.exports)
+			.then(resolve)
+			.catch(resolve);
+	}
+`;
+
+export interface ExportDescription {
+	localName: string;
+	identifier?: string;
+}
+
+export interface ImportDescription {
+	source: string;
+	name: string;
+	module: Module | ExternalModule | null;
+}
+
+export default class WasmModule {
+	id: string;
+	graph: Graph;
+	code: Buffer;
+
+	scope: ModuleScope;
+
+	ast: WebAssemblyJSAst.Program;
```

Comment:
We could provide the same AST manipulation logic to plugins as with JS.

Reply:
Awesome!

Reply:
Forgot to mention that we can enable AST diffing + wasm module manipulation using this API. Which could be useful for plugin (mutable global polyfilling for ex :100: ).